### PR TITLE
ci-operator/templates/openshift/installer/cluster-launch-installer-upi*: Switch on BRANCH for RHCOS AMI

### PR DESCRIPTION
**Pull request by wking**
As discussed in dc26075230 (#6192), we want a way to switch on the z stream when selecting the RHCOS AMI.  We amight also need this if other parts of the recommended user-provided-infrastructure flow diverge between z streams, although at the moment the AMI is the only divergence.

In this commit, I'm sourcing my AMIs with:

```console
$ git cat-file -p origin/release-4.1:data/data/rhcos.json | jq -r '.amis["us-east-1"].hvm'
ami-046fe691f52a953f9
$ git cat-file -p origin/release-4.2:data/data/rhcos.json | jq -r '.amis["us-east-1"].hvm'
ami-01e7fdcb66157b224
$ git cat-file -p origin/release-4.3:data/data/rhcos.json | jq -r '.amis["us-east-1"].hvm'
ami-014ce8846db8b463d
$ git cat-file -p origin/master:data/data/rhcos.json | jq -r '.amis["us-east-1"].hvm'
ami-014ce8846db8b463d
```

as the installer stood at 2019-12-04T23:30Z.  Comparing with the user-facing docs:

```console
$ git --no-pager grep -hA1 us-east-1 origin/enterprise-4.1:modules/installation-aws-user-infra-rhcos-ami.adoc
|`us-east-1`
|`ami-046fe691f52a953f9`
$ git --no-pager grep -hA1 us-east-1 origin/enterprise-4.2:modules/installation-aws-user-infra-rhcos-ami.adoc
|`us-east-1`
|`ami-01e7fdcb66157b224`
$ git --no-pager grep -hA1 us-east-1 origin/enterprise-4.3:modules/installation-aws-user-infra-rhcos-ami.adoc
|`us-east-1`
|`ami-01e7fdcb66157b224`
$ git --no-pager grep -hA1 us-east-1 origin/enterprise-4.4:modules/installation-aws-user-infra-rhcos-ami.adoc
fatal: Invalid object name 'origin/enterprise-4.4'.
$ git --no-pager grep -hA1 us-east-1 origin/master:modules/installation-aws-user-infra-rhcos-ami.adoc
|`us-east-1`
|`ami-01e7fdcb66157b224`
```

So we currently diverge for 4.3 and master, where I have openshift/openshift-docs#18396 and openshift/openshift-docs#18395 open to close the gaps.

The BRANCH environment variable has been around since 22e30eac2b (#3649).  I added it to all jobs based on the UPI template where it was missing (it was already on a number of those jobs).

### CODE REVIEW

**Comment by wking** [comment]
Classifying the early errors:

```console
$ curl -s 'https://ci-search-ci-search-next.svc.ci.openshift.org/search?name=rehearse-6248&search=error:%20build%20error:%20no%20such%20image&search=could%20not%20resolve%20base%20image.*not%20found&search=syntax%20error%20near%20unexpected%20token' | jq -r '. | to_entries[] | .key + " " + (.value | keys | join(" "))'
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.2-e2e-aws-upi/1 error: build error: no such image
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.2-e2e-vsphere/1 error: build error: no such image
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.4-e2e-aws-upi/1 error: build error: no such image
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.4-e2e-gcp-upi/1 error: build error: no such image
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.4-e2e-vsphere/1 error: build error: no such image
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-aws-upi/1 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-gcp-upi/1 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-vsphere/1 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-machine-config-operator-release-4.4-e2e-vsphere/1 syntax error near unexpected token
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-machine-config-operator-release-4.5-e2e-vsphere/1 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.5-e2e-vsphere/1 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.5-launch-vsphere/1 could not resolve base image.*not found
```

None of which seem related to this PR.

**Comment by wking** [comment]
Never mind, [the][1]:

    /bin/sh: -c: line 871: syntax error near unexpected token `)'

was because I'd forgotten closing `;;`.  Fixed with fd2b0e18c -> 4972b87df.

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-machine-config-operator-release-4.4-e2e-vsphere/1

**Comment by wking** [comment]
Early failures in the second round of CI:

```console
$ curl -s 'https://ci-search-ci-search-next.svc.ci.openshift.org/search?name=rehearse-6248-&search=error:+build+error:+no+such+image&search=could+not+resolve+base+image.*not+found&search=syntax+error+near+unexpected+token&search=Terraform+does+not+automatically+rollback+in+the+face+of+errors' | jq -r '. | to_entries[] | .key + " " + (.value | keys | join(" "))' | grep /2
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.2-e2e-vsphere/2 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-aws-upi/2 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-gcp-upi/2 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-vsphere/2 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-machine-config-operator-release-4.3-e2e-vsphere/2 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-machine-config-operator-release-4.5-e2e-vsphere/2 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.5-e2e-vsphere/2 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.5-launch-vsphere/2 could not resolve base image.*not found
```

Now none of the failures look related to this PR.

**Comment by wking** [comment]
[4.1 AWS UPI][1]:

```
An error occurred (ValidationError) when calling the CreateStack operation: Parameters: [WorkerSubnet] must have values
```

This is probably a place where our UPI flow has diverged, because openshift/installer#1794 was never backported to 4.1.

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.1-e2e-aws-upi/2

**Comment by wking** [comment]
Oops.  Fumbled the first attempt at `WorkerSubnet` for 4.1 and `Subnet` for 4.2+.  Pushed 03c0c4907 -> 545d327e7 with a fix.  Results from round 3:

```console
$ curl -s 'https://ci-search-ci-search-next.svc.ci.openshift.org/search?name=rehearse-6248-&search=error:+build+error:+no+such+image&search=could+not+resolve+base+image.*not+found&search=syntax+error+near+unexpected+token&search=Terraform+does+not+automatically+rollback+in+the+face+of+errors&search=Subnet.*must+have+values&search=add-iam-policy-binding.*is+the+subject+of+a+conflict&search=Failing+tests' | jq -r '. | to_entries[] | .key + " " + (.value | keys | join(" "))' | grep /3
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.1-e2e-aws-upi/3 Failing tests
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.1-e2e-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.2-e2e-aws-upi/3 Subnet.*must have values
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.2-e2e-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.3-e2e-aws-upi/3 Subnet.*must have values
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.3-e2e-gcp-upi/3 add-iam-policy-binding.*is the subject of a conflict
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.3-e2e-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.4-e2e-aws-upi/3 Subnet.*must have values
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.4-e2e-vsphere/3 Failing tests
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-aws-upi/3 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-gcp-upi/3 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-installer-release-4.5-e2e-vsphere/3 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-machine-config-operator-release-4.3-e2e-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-machine-config-operator-release-4.4-e2e-vsphere/3 Failing tests
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-machine-config-operator-release-4.5-e2e-vsphere/3 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.1-e2e-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.1-launch-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.2-e2e-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.3-e2e-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.3-launch-vsphere/3 Terraform does not automatically rollback in the face of errors
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.5-e2e-vsphere/3 could not resolve base image.*not found
https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.5-launch-vsphere/3 could not resolve base image.*not found
```

**Comment by wking** [comment]
Still [one 4.4 vSphere job][1] from round 4 in flight, but here are the rest of them (putting the regexp before the job name so we can collect by regexp):

```console
$ curl -s 'https://ci-search-ci-search-next.svc.ci.openshift.org/search?name=rehearse-6248-&search=error:+build+error:+no+such+image&search=could+not+resolve+base+image.*not+found&search=syntax+error+near+unexpected+token&search=Terraform+does+not+automatically+rollback+in+the+face+of+errors&search=Subnet.*must+have+values&search=add-iam-policy-binding.*is+the+subject+of+a+conflict&search=Failing+tests&search=Waiter+StackDeleteComplete+failed:+Waiter+encountered+a+terminal+failure+state&search=failed+to+initialize+the+cluster' | jq -r '. | to_entries[] | .key + " " + (.value | keys | join(" "))' | sed -n 's|.*ci-openshift-\([^ ]*/4\) \(.*\)|\2\t\1|p' | sort
could not resolve base image.*not found	installer-release-4.5-e2e-aws-upi/4
could not resolve base image.*not found	installer-release-4.5-e2e-gcp-upi/4
could not resolve base image.*not found	installer-release-4.5-e2e-vsphere/4
could not resolve base image.*not found	machine-config-operator-release-4.5-e2e-vsphere/4
could not resolve base image.*not found	origin-release-4.5-e2e-vsphere/4
could not resolve base image.*not found	origin-release-4.5-launch-vsphere/4
failed to initialize the cluster	installer-release-4.4-e2e-gcp-upi/4
Failing tests	installer-release-4.1-e2e-aws-upi/4
Failing tests	installer-release-4.2-e2e-aws-upi/4
Failing tests	installer-release-4.4-e2e-aws-upi/4
Failing tests	installer-release-4.4-e2e-vsphere/4
Failing tests	machine-config-operator-release-4.4-e2e-vsphere/4
Failing tests	origin-release-4.4-e2e-vsphere/4
Terraform does not automatically rollback in the face of errors	installer-release-4.1-e2e-vsphere/4
Terraform does not automatically rollback in the face of errors	installer-release-4.2-e2e-vsphere/4
Terraform does not automatically rollback in the face of errors	installer-release-4.3-e2e-vsphere/4
Terraform does not automatically rollback in the face of errors	machine-config-operator-release-4.3-e2e-vsphere/4
Terraform does not automatically rollback in the face of errors	origin-release-4.1-e2e-vsphere/4
Terraform does not automatically rollback in the face of errors	origin-release-4.1-launch-vsphere/4
Terraform does not automatically rollback in the face of errors	origin-release-4.2-e2e-vsphere/4
Terraform does not automatically rollback in the face of errors	origin-release-4.2-launch-vsphere/4
Terraform does not automatically rollback in the face of errors	origin-release-4.3-e2e-vsphere/4
Terraform does not automatically rollback in the face of errors	origin-release-4.3-launch-vsphere/4
Waiter StackDeleteComplete failed: Waiter encountered a terminal failure state	installer-release-4.3-e2e-aws-upi/4
```

None of those seem to be related to this PR, so I think we're good to go here if we want this direction.

And taking another stab at motivating each of the `BRANCH` cases:

* Using 4.2 RHCOS avoids bootstrap trouble using 4.3 AMIs.  For example, [this 4.2 proxy job][2] had [a log tarball][3] whose `./bootstrap/journals/kubelet.log` showed crashlooping on:

        Dec 04 16:06:14 ip-10-0-4-179 hyperkube[2242]: F1204 16:06:14.652676    2242 server.go:154] unknown flag: --allow-privileged

    The installer stopped setting `--allow-privileged` in openshift/installer#2459, but that was never bacported to 4.2-.  Using a 4.2 AMI for 4.2 UPI avoids this issue.  You could probably get away with using a 4.2 AMI for 4.1 UPI if you wanted, although with this PR I'm using a 4.1 AMI for 4.1 UPI.

    You could also avoid this issue by treating the bootimage (AMI) as a pivot-only image and pivoting the bootstrap machine to the release-references machine-os-content (like we do for cluster machines).  That's getting kicked around in openshift/installer#2542, but it's not what the installer does today (and certainly not in 4.1/4.2 today).

* Using the 4.2+ `Subnet` CloudFormation template property breaks 4.1, where the template expects `WorkerSubnet` because openshift/installer#1794 landed in 4.2+ but was never backported to 4.1.  Using `BRANCH` to select the right parameter key works.  You could also avoid this by backporting the parameter rename, but that would break users who had coded flows around the current 4.1 docs.  On the other hand, we do say the CloudFormation templates are guidelines, and we never suggest folks pull them straight from the installer repo, so there's some wiggle room here.  We could backport openshift/installer#1794 to 4.1.z, and leave the docs alone, and just live with the divergence.  Or we could update the docs and live with the divergence from folks who had copied out the previous templates into some external workflow.

So there are a number of ways we could do this, and while I like the `BRANCH` approach, I'm fine taking one of the alternatives instead.  If we end up deciding we want something like "teach 4.1.z to pivot the bootstrap machine" that will take a while to write up and backport to 4.1.z, I think we need to land *something* in the interim to unblock 4.1 and 4.2 UPI (or revert #6192 and re-break 4.3 UPI).  Thoughts?  CC @abhinavdahiya 

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/rehearse-6248-pull-ci-openshift-origin-release-4.4-launch-vsphere/4
[2]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/logs/release-openshift-ocp-installer-e2e-aws-proxy-4.2/220
[3]: https://gcsweb-ci.svc.ci.openshift.org/gcs/origin-ci-test/logs/release-openshift-ocp-installer-e2e-aws-proxy-4.2/220/artifacts/e2e-aws-proxy/installer/

File: ci-operator/templates/openshift/installer/cluster-launch-installer-upi-src.yaml [hunk 1]
```diff
@@ -575,7 +576,13 @@ objects:
 
         # begin bootstrapping
         if [[ "${CLUSTER_TYPE}" == "aws" ]]; then
-          RHCOS_AMI=ami-014ce8846db8b463d  # FIXME: assumes AWS_REGION is us-east-1
+          case "${BRANCH}" in  # FIXME: assumes AWS_REGION is us-east-1
+          4.1) RHCOS_AMI=ami-046fe691f52a953f9;;
+          4.2) RHCOS_AMI=ami-01e7fdcb66157b224;;
+          4.3) RHCOS_AMI=ami-014ce8846db8b463d;;
+          *) RHCOS_AMI=ami-014ce8846db8b463d;;
```
Comment (cgwalters) [inline ci-operator/templates/openshift/installer/cluster-launch-installer-upi-src.yaml]: This is OK but I would prefer automating https://github.com/openshift/installer/pull/2092#issuecomment-522663307

**Comment by wking** [comment]
/hold

Looks like [generated-config][1] is sad:

```
 diff -Naupr hack/../ci-operator/jobs/openshift/installer/openshift-installer-release-4.1-presubmits.yaml /tmp/tmp.So5Op8vb7e/ci-operator/jobs/openshift/installer/openshift-installer-release-4.1-presubmits.yaml
--- hack/../ci-operator/jobs/openshift/installer/openshift-installer-release-4.1-presubmits.yaml	2019-12-05 22:23:16.504454851 +0000
+++ /tmp/tmp.So5Op8vb7e/ci-operator/jobs/openshift/installer/openshift-installer-release-4.1-presubmits.yaml	2019-12-05 22:23:39.982709279 +0000
@@ -236,8 +236,6 @@ presubmits:
         command:
         - ci-operator
         env:
-        - name: BRANCH
-          value: "4.1"
         - name: CLUSTER_TYPE
           value: aws
         - name: CONFIG_SPEC 
...
```

Not clear on why it doesn't like the environment variable...

[1]: https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_release/6248/pull-ci-openshift-release-master-generated-config/12911

**Comment by wking** [comment]
Talked this over with @stevekuznetsov and @droslean out of band.  Takeaway is that this sort of conditional logic is not supported.  I'm going to wait for the multi-step stuff (e.g. openshift/ci-tools#319) and take another run at this then.

/close

**Comment by wking** [comment]
I've filed https://bugzilla.redhat.com/show_bug.cgi?id=1782062 and https://bugzilla.redhat.com/show_bug.cgi?id=1782063 to track the failures until we get a path to fixing them.

**Pull request closed without merging**